examples/hacspec-gimli/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+        let ct_block_padded = Block::new();
+        let ct_block_padded = ct_block_padded.update_start(&ct_block);
+        let msg_block = ct_block_padded ^ key_block;
+
+        // XXX: pretty ugly. Zero pad the last block if it is padded.
+        let mut msg_block = Block::from_slice_range(&msg_block, 0..block_len);
+
+        // Slice_range cuts off the msg_block to the actual length.
+        message = message.set_chunk(rate, i, &msg_block.slice_range(0..block_len));
+        if i == num_chunks - 1 {
+            // XXX: this is in the code but I don't see it in the spec.
+            msg_block[block_len] = msg_block[block_len] ^ U8(1u8);
+            s[11] = s[11] ^ U32(0x01000000u32); // s_2,3
+        }
+        s = absorb_block(msg_block, s);
+    }
+
+    (s, message)
+}
+
+// XXX: These two functions should maybe get a helper in the library.
+pub fn nonce_to_u32s(nonce: Nonce) -> Seq<U32> {
+    let mut uints = Seq::<U32>::new(4);
+    uints[0] = U32_from_le_bytes(U32Word::from_slice_range(&nonce, 0..4));
+    uints[1] = U32_from_le_bytes(U32Word::from_slice_range(&nonce, 4..8));
+    uints[2] = U32_from_le_bytes(U32Word::from_slice_range(&nonce, 8..12));
+    uints[3] = U32_from_le_bytes(U32Word::from_slice_range(&nonce, 12..16));
+    uints
+}
+
+pub fn key_to_u32s(key: Key) -> Seq<U32> {
+    let mut uints = Seq::<U32>::new(8);
+    uints[0] = U32_from_le_bytes(U32Word::from_slice_range(&key, 0..4));
+    uints[1] = U32_from_le_bytes(U32Word::from_slice_range(&key, 4..8));
+    uints[2] = U32_from_le_bytes(U32Word::from_slice_range(&key, 8..12));
+    uints[3] = U32_from_le_bytes(U32Word::from_slice_range(&key, 12..16));
+    uints[4] = U32_from_le_bytes(U32Word::from_slice_range(&key, 16..20));
+    uints[5] = U32_from_le_bytes(U32Word::from_slice_range(&key, 20..24));
+    uints[6] = U32_from_le_bytes(U32Word::from_slice_range(&key, 24..28));
+    uints[7] = U32_from_le_bytes(U32Word::from_slice_range(&key, 28..32));
+    uints
+}
+
+pub fn gimli_aead_encrypt(
+    message: &ByteSeq,
+    ad: &ByteSeq,
+    nonce: Nonce,
+    key: Key,
+) -> (ByteSeq, Tag) {
+    // Add nonce and key to state
+    let s = State::from_seq(&nonce_to_u32s(nonce).concat(&key_to_u32s(key)));
+    let s = gimli(s);
+
+    let s = process_ad(ad, s);
+    let (s, ciphertext) = process_msg(message, s);
+
+    let tag = squeeze_block(s);
+    // XXX: not great
+    let mut public_tag = Tag::new();
+    for i in 0..Tag::length() {
+        public_tag[i] = tag[i].declassify();
+    }
+    (ciphertext, public_tag)
+}
+
+pub fn gimli_aead_decrypt(
+    ciphertext: &ByteSeq,
+    ad: &ByteSeq,
+    tag: Tag,
+    nonce: Nonce,
+    key: Key,
+) -> ByteSeq {
+    // Add nonce and key to state
+    let s = State::from_seq(&nonce_to_u32s(nonce).concat(&key_to_u32s(key)));
+    let s = gimli(s);
+
+    let s = process_ad(ad, s);
+    let (s, message) = process_ct(ciphertext, s);
+
+    let my_tag = squeeze_block(s);
+
+    // XXX: not great
```

Comment:
It would be doable to add a `seq_declassify` function in the lib that does that I think. Its type would have to be annotated though so you would have to do `my_public_tag = seq_declassify::<U8>(my_tag)`.

Reply:
Yeah I wonder if we should add something like this. Making values like these public appears to be a relatively common thing. I actually settled with using secret bytes for the tag as well.